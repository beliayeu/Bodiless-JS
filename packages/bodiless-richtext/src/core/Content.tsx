/**
 * Copyright Â© 2019 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Editable, DefaultElement, DefaultLeaf } from 'slate-react';
import { useSlateContext } from './SlateEditorContext';
import type { EditorContext } from '../Type';

type EditableProps = EditorContext['editorProps'];

const withWrapper = WrapperComponent => Component => ({children, ...rest}) => (
  <WrapperComponent {...rest}>
    <Component>{children}</Component>
  </WrapperComponent>
)


const renderLeaf = (props) => {
  const { leaf } = props;
  const editorContext = useSlateContext();
  const plugins = editorContext?.editorProps!.plugins;
  let renderLeaf$ = DefaultLeaf;
  plugins.forEach(plugin => {
    if (plugin.hasOwnProperty('renderLeaf') && leaf[plugin.type]) {
      renderLeaf$ = withWrapper(plugin.renderLeaf)(renderLeaf$);
    }
  });
  return renderLeaf$(props);
}

const renderElement = (props) => {
  const { element } = props;
  const editorContext = useSlateContext();
  const plugins = editorContext?.editorProps!.plugins;
  let renderElement$ = DefaultElement;
  plugins.forEach(plugin => {
    if (plugin.hasOwnProperty('renderElement') && element.type === plugin.type) {
      renderElement$ = plugin.renderElement;
    }
  });
  return renderElement$(props);
}

const Content = (props: EditableProps) => {
  const editorContext = useSlateContext();
  return (
    <Editable
      {...props}
      { ...editorContext!.editorProps}
      renderLeaf={renderLeaf}
      renderElement={renderElement}
    />
  );
};

export default Content;
