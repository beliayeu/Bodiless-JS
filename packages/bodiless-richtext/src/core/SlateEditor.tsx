/**
 * Copyright Â© 2019 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, {
  useState,
  Fragment,
  ComponentType,
} from 'react';
import SlateEditorContext from './SlateEditorContext';
import '@material/react-material-icon/dist/material-icon.css';
import { EditorOnChange } from '../Type';

type Props = {
  initialValue: object;
} & EditorProps;

const withSlateEditor = <P extends object> (Component:ComponentType<P>) => (props:P & Props) => {
  const {
    initialValue, value, onChange, placeholder, plugins = [], ...rest
  } = props;
  // It is important to keep track of internal activeValue
  // state in case outer activeValue is not provided.
  // Value is used in plugins and buttons before Content is mounted and its activeValue is obtained.
  const [localValueState, setLocalValue] = useState<object>(
    value || initialValue
  );
  const internalOnChange: EditorOnChange = change => {
    const { value: valueFromChange } = change;
    if (typeof onChange === 'function') {
      onChange(change);
    } else if (typeof value !== 'undefined') {
      setLocalValue(valueFromChange);
    }

    return change;
  };

  const editorContextValue = {
    value: value || localValueState,
    editorProps: {
      ...rest,
      plugins,
      onChange: internalOnChange,
      value: value || localValueState,
    },
  };

  return (
    <SlateEditorContext.Provider value={editorContextValue}>
      <Component {...rest as P} />
    </SlateEditorContext.Provider>
  );
};
const SlateEditor = withSlateEditor(Fragment);

export default SlateEditor;
export { withSlateEditor };
