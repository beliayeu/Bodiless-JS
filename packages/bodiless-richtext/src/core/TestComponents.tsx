import React, { Ref, PropsWithChildren } from 'react'
import ReactDOM from 'react-dom'
import { cx, css } from 'emotion'

import { Editable, withReact, useSlate, Slate } from 'slate-react'
import { Editor, Transforms, createEditor, Node } from 'slate'

interface BaseProps {
  className: string
  [key: string]: unknown
}
type OrNull<T> = T | null

export const Button = React.forwardRef(
  (
    {
      className,
      active,
      reversed,
      ...props
    }: PropsWithChildren<
      {
        active: boolean
        reversed: boolean
      } & BaseProps
    >,
    ref: Ref<OrNull<HTMLSpanElement>>
  ) => (
      <span
        {...props}
        ref={ref}
        className={cx(
          className,
          css`
          cursor: pointer;
          color: ${reversed
              ? active
                ? 'white'
                : '#aaa'
              : active
                ? 'black'
                : '#ccc'};
        `
        )}
      />
    )
)

export const EditorValue = React.forwardRef(
  (
    {
      className,
      value,
      ...props
    }: PropsWithChildren<
      {
        value: any
      } & BaseProps
    >,
    ref: Ref<OrNull<null>>
  ) => {
    const textLines = value.document.nodes
      .map(node => node.text)
      .toArray()
      .join('\n')
    return (
      <div
        ref={ref}
        {...props}
        className={cx(
          className,
          css`
            margin: 30px -20px 0;
          `
        )}
      >
        <div
          className={css`
            font-size: 14px;
            padding: 5px 20px;
            color: #404040;
            border-top: 2px solid #eeeeee;
            background: #f8f8f8;
          `}
        >
          Slate's value as text
        </div>
        <div
          className={css`
            color: #404040;
            font: 12px monospace;
            white-space: pre-wrap;
            padding: 10px 20px;
            div {
              margin: 0 0 0.5em;
            }
          `}
        >
          {textLines}
        </div>
      </div>
    )
  }
)

export const Icon = React.forwardRef(
  (
    { className, ...props }: PropsWithChildren<BaseProps>,
    ref: Ref<OrNull<HTMLSpanElement>>
  ) => (
      <span
        {...props}
        ref={ref}
        className={cx(
          'material-icons',
          className,
          css`
          font-size: 18px;
          vertical-align: text-bottom;
        `
        )}
      />
    )
)

export const Instruction = React.forwardRef(
  (
    { className, ...props }: PropsWithChildren<BaseProps>,
    ref: Ref<OrNull<HTMLDivElement>>
  ) => (
      <div
        {...props}
        ref={ref}
        className={cx(
          className,
          css`
          white-space: pre-wrap;
          margin: 0 -20px 10px;
          padding: 10px 20px;
          font-size: 14px;
          background: #f8f8e8;
        `
        )}
      />
    )
)

export const Menu = React.forwardRef(
  (
    { className, ...props }: PropsWithChildren<BaseProps>,
    ref: Ref<OrNull<HTMLDivElement>>
  ) => (
      <div
        {...props}
        ref={ref}
        className={cx(
          className,
          css`
          & > * {
            display: inline-block;
          }

          & > * + * {
            margin-left: 15px;
          }
        `
        )}
      />
    )
)

export const Portal = ({ children }) => {
  return ReactDOM.createPortal(children, document.body)
}

export const Toolbar = React.forwardRef(
  (
    { className, ...props }: PropsWithChildren<BaseProps>,
    ref: Ref<OrNull<HTMLDivElement>>
  ) => (
      <Menu
        {...props}
        ref={ref}
        className={cx(
          className,
          css`
          position: relative;
          padding: 1px 18px 17px;
          margin: 0 -20px;
          border-bottom: 2px solid #eee;
          margin-bottom: 20px;
        `
        )}
      />
    )
)


const LIST_TYPES = ['numbered-list', 'bulleted-list']


const isBlockActive = (editor, format) => {
  const [match] = Editor.nodes(editor, {
    match: n => n.type === format,
  })

  return !!match
}


const toggleBlock = (editor, format) => {
  const isActive = isBlockActive(editor, format)
  const isList = LIST_TYPES.includes(format)

  Transforms.unwrapNodes(editor, {
    match: n => LIST_TYPES.includes(n.type as string),
    split: true,
  })

  Transforms.setNodes(editor, {
    type: isActive ? 'paragraph' : isList ? 'list-item' : format,
  })

  if (!isActive && isList) {
    const block = { type: format, children: [] }
    Transforms.wrapNodes(editor, block)
  }
}

export const BlockButton = ({ format, icon }) => {
  const editor = useSlate()
  return (
    <Button
      active={isBlockActive(editor, format)}
      onMouseDown={event => {
        event.preventDefault()
        toggleBlock(editor, format)
      }}
    >
      <Icon>{icon}</Icon>
    </Button>
  )
}

const isMarkActive = (editor, format) => {
  const marks = Editor.marks(editor)
  return marks ? marks[format] === true : false
}

const toggleMark = (editor, format) => {
  const isActive = isMarkActive(editor, format)

  if (isActive) {
    Editor.removeMark(editor, format)
  } else {
    Editor.addMark(editor, format, true)
  }
}

export const MarkButton = ({ format, icon }) => {
  const editor = useSlate()
  return (
    <Button
      active={isMarkActive(editor, format)}
      onMouseDown={event => {
        event.preventDefault()
        toggleMark(editor, format)
      }}
    >
      <Icon>{icon}</Icon>
    </Button>
  )
}
