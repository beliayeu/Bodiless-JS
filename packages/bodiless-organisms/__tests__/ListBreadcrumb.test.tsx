/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { mount } from 'enzyme';
import { withDefaultContent } from '@bodiless/core';
import {
  asBodilessList,
  asBodilessLink,
  asSubList,
  withSimpleSubListDesign,
  ifActiveBreadcrumb,
} from '@bodiless/components';
import {
  withDesign,
  replaceWith,
  asComponent,
  addProps,
} from '@bodiless/fclasses';
import { flowRight, identity } from 'lodash';
// ToDo: remove this
import { html_beautify } from 'js-beautify';

import asBreadcrumbsClean, {
  withBreadcrumbSeparator,
  withBreadcrumbDesign,
} from '../src/components/ListBreadcrumb';

const { DefaultContentNode } = require('@bodiless/core');

const withSubList = (depth: number) => withSimpleSubListDesign(depth)(
  // @ts-ignore ToDo: resolve types
  flowRight(
    withDesign({
      Title: replaceWith(
        asBodilessLink('link')('a')
      ),
    }),
    asSubList,
  ),
);

const setPagePath = (pagePath: string) => {
  Object.defineProperty(DefaultContentNode.prototype, 'pagePath', {
    value: pagePath,
    writable: false
  });
};

const createBreadcrumbComponent = ({
  // ToDo: we can determine depth automatically based on content
  depth = 1,
  content = {},
}) => flowRight(
  asBreadcrumbsClean({ depth }),
  withDefaultContent(content),
  depth > 1 ? withSubList(depth - 1) : identity,
  withDesign({
    Title: replaceWith(
      asBodilessLink('link')('a')
    ),
  }),
  asBodilessList('testList')
)('ul');

describe('asBreadcrumbsClean', () => {
  it('creates breadcrumbs for basic 1-level list', () => {
    setPagePath('/products');
    const Breadcrumb = createBreadcrumbComponent({
      depth: 1,
      content: {
        testList: {
          "items": [
            "home",
            "products",
            "articles",
          ],
        },
        testList$home$link: {
          'href': '/'
        },
        testList$products$link: {
          'href': '/products'
        },
        testList$articles$link: {
          'href': '/articles'
        },
      }      
    });
    const wrapper = mount(<Breadcrumb />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it ('creates breadcrumbs for basic 2-level list', () => {
    setPagePath('/products/productA');
    const Breadcrumb = createBreadcrumbComponent({
      depth: 2,
      content: {
        testList: {
          "items": [
            "home",
            "products",
            "articles",
          ],
        },
        testList$home$link: {
          'href': '/'
        },
        testList$products$link: {
          'href': '/products'
        },
        testList$products$sublist: {
          "items": [
            "productA",
            "productB",
          ],
        },
        testList$products$sublist$productA$link: {
          'href': '/products/productA'
        },
        testList$products$sublist$productB$link: {
          'href': '/products/productB'
        },
        testList$articles$link: {
          'href': '/articles'
        },
        testList$articles$sublist: {
          "items": [
            "articleA",
          ],
        },
        testList$articles$sublist$articleA$link: {
          'href': '/articles/articleA'
        },
      }
    });
    const wrapper = mount(<Breadcrumb />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it ('creates breadcrumbs for basic 3-level list', () => {
    setPagePath('/articles/news/newsA');
    const Breadcrumb = createBreadcrumbComponent({
      depth: 3,
      content: {
        testList: {
          "items": [
            "home",
            "articles",
          ],
        },
        testList$home$link: {
          'href': '/'
        },
        testList$articles$link: {
          'href': '/articles'
        },
        testList$articles$sublist: {
          "items": [
            "news",
          ],
        },
        testList$articles$sublist$news$link: {
          'href': '/articles/news'
        },
        testList$articles$sublist$news$sublist: {
          "items": [
            "newsA",
          ],
        },
        testList$articles$sublist$news$sublist$newsA$link: {
          'href': '/articles/news/newsA'
        },
      }
    });
    const wrapper = mount(<Breadcrumb />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('creates breadcrumbs when child is active and parent is not subpath of child', () => {
    setPagePath('/productA');
    const Breadcrumb = createBreadcrumbComponent({
      depth: 2,
      content: {
        testList: {
          "items": [
            "home",
            "products",
          ],
        },
        testList$home$link: {
          'href': '/'
        },
        testList$products$link: {
          'href': '/products'
        },
        testList$products$sublist: {
          "items": [
            "productA",
          ],
        },
        testList$products$sublist$productA$link: {
          'href': '/productA'
        },
      }
    });
    const wrapper = mount(<Breadcrumb />);
    console.log(html_beautify(wrapper.html()));
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe('withBreadcrumbSeparator', () => {
  it('can add a separator to breadcrumbs', () => {
    setPagePath('/products/productA');
    const Breadcrumb = createBreadcrumbComponent({
      depth: 2,
      content: {
        testList: {
          "items": [
            "home",
            "products",
            "articles",
          ],
        },
        testList$home$link: {
          'href': '/'
        },
        testList$products$link: {
          'href': '/products'
        },
        testList$products$sublist: {
          "items": [
            "productA",
            "productB",
          ],
        },
        testList$products$sublist$productA$link: {
          'href': '/products/productA'
        },
        testList$products$sublist$productB$link: {
          'href': '/products/productB'
        },
        testList$articles$link: {
          'href': '/articles'
        },
        testList$articles$sublist: {
          "items": [
            "articleA",
          ],
        },
        testList$articles$sublist$articleA$link: {
          'href': '/articles/articleA'
        },
      }
    });
    // @ts-ignore
    const BreadcrumbWithSeparator = withDesign({
      Item: flowRight(
        withDesign({
          Separator: replaceWith(() => <div id="parentSeparator" />),
        }),
      ),
    })(Breadcrumb);
    const wrapper = mount(<BreadcrumbWithSeparator />);
    console.log(html_beautify(wrapper.html()));
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe('withBreadcrumbDesign', () => {
  it('allows designing breadcrumb components', () => {
    setPagePath('/products/productA');
    const Breadcrumb = createBreadcrumbComponent({
      depth: 2,
      content: {
        testList: {
          "items": [
            "home",
            "products",
            "articles",
          ],
        },
        testList$home$link: {
          'href': '/'
        },
        testList$products$link: {
          'href': '/products'
        },
        testList$products$sublist: {
          "items": [
            "productA",
            "productB",
          ],
        },
        testList$products$sublist$productA$link: {
          'href': '/products/productA'
        },
        testList$products$sublist$productB$link: {
          'href': '/products/productB'
        },
        testList$articles$link: {
          'href': '/articles'
        },
        testList$articles$sublist: {
          "items": [
            "articleA",
          ],
        },
        testList$articles$sublist$articleA$link: {
          'href': '/articles/articleA'
        },
      }
    });
    const DesignedBreadcrumb = flowRight(
      withBreadcrumbDesign({
        Title: addProps({ className: 'breadcrumb-title'}),
        Item: addProps({ className: 'breadcrumb-item'}),
      }),
    )(Breadcrumb);
    const wrapper = mount(<DesignedBreadcrumb />);
    expect(wrapper.html()).toMatchSnapshot();
  });
});