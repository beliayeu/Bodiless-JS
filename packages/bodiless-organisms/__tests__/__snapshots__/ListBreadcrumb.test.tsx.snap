// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asBreadcrumbsClean creates breadcrumbs for basic 1-level list 1`] = `"<ul><li><a href=\\"/products\\"></a></li></ul>"`;

exports[`asBreadcrumbsClean creates breadcrumbs for basic 2-level list 1`] = `"<ul><li><a href=\\"/products\\"></a></li><li><a href=\\"/products/productA\\"></a></li></ul>"`;

exports[`asBreadcrumbsClean creates breadcrumbs for basic 3-level list 1`] = `"<ul><li><a href=\\"/articles\\"></a></li><li><a href=\\"/articles/news\\"></a></li><li><a href=\\"/articles/news/newsA\\"></a></li></ul>"`;

exports[`asBreadcrumbsClean creates breadcrumbs when child is active and parent is not subpath of child 1`] = `"<ul><li><a href=\\"/products\\"></a></li><li design=\\"[object Object]\\"><a href=\\"/productA\\"></a></li></ul>"`;

exports[`withBreadcrumbDesign allows designing breadcrumb components 1`] = `"<ul><li label=\\"breadcrumb-item\\"><a href=\\"/products\\" class=\\"breadcrumb-title\\"></a></li><li label=\\"breadcrumb-item\\" design=\\"[object Object]\\"><a href=\\"/products/productA\\" class=\\"breadcrumb-title\\"></a></li></ul>"`;

exports[`withBreadcrumbSeparator can add a separator to breadcrumbs 1`] = `"<ul><li><a href=\\"/products\\"></a></li><li design=\\"[object Object]\\"><a href=\\"/products/productA\\"></a></li><span id=\\"separator\\"></span><span id=\\"separator\\"></span></ul>"`;
